python/zfp.pyx

```diff
@@ -0,0 +1,210 @@
+
+zfp_to_dtype_map = {
+    zfp_type_int32: np.int32,
+    zfp_type_int64: np.int64,
+    zfp_type_float: np.float32,
+    zfp_type_double: np.float64,
+}
+cpdef ztype_to_dtype(zfp_type ztype):
+    try:
+        return zfp_to_dtype_map[ztype]
+    except KeyError:
+        # TODO: is this the correct type?
+        print "Using np.bytes_"
+        return np.bytes_
+
+cdef size_t sizeof_ztype(zfp_type ztype):
+    if ztype == zfp_type_int32:
+        return sizeof(signed int)
+    elif ztype == zfp_type_int64:
+        return sizeof(signed long long)
+    elif ztype == zfp_type_float:
+        return sizeof(float)
+    elif ztype == zfp_type_double:
+        return sizeof(double)
+    else:
+        return -1
+
+cdef zfp_field* _init_field(np.ndarray arr):
+    shape = arr.shape
+    cdef int ndim = arr.ndim
+    cdef zfp_type ztype = dtype_to_ztype(arr.dtype)
+    cdef zfp_field* field
+    cdef void* pointer = arr.data
+
+    strides = [int(x) / arr.itemsize for x in arr.strides[:ndim]]
+
+    if ndim == 1:
+        field = zfp_field_1d(pointer, ztype, shape[0])
+        zfp_field_set_stride_1d(field, strides[0])
+    elif ndim == 2:
+        field = zfp_field_2d(pointer, ztype, shape[0], shape[1])
+        zfp_field_set_stride_2d(field, strides[0], strides[1])
+    elif ndim == 3:
+        field = zfp_field_3d(pointer, ztype, shape[0], shape[1], shape[2])
+        zfp_field_set_stride_3d(field, strides[0], strides[1], strides[2])
+    elif ndim == 4:
+        field = zfp_field_4d(pointer, ztype, shape[0], shape[1], shape[2], shape[3])
+        zfp_field_set_stride_4d(field, strides[0], strides[1], strides[2], strides[3])
+    else:
+        raise RuntimeError("Greater than 4 dimensions not supported")
+
+    return field
+
+@cython.final
+cdef class Memory:
+    cdef void* data
+    def __cinit__(self, size_t size):
+        self.data = malloc(size)
+        if self.data == NULL:
+            raise MemoryError()
+    cdef void* __enter__(self):
+        return self.data
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        free(self.data)
+
+cpdef bytes compress_numpy(np.ndarray arr, double tolerance = -1,
+                           double rate = -1, int precision = -1,
+                           write_header=True):
+    # Input validation
+    if arr is None:
+        raise TypeError("Input array cannot be None")
+    num_params_set = sum([1 for x in [tolerance, rate, precision] if x > -1])
+    if num_params_set == 0:
+        raise ValueError("Either tolerance, rate, or precision must be set")
+    elif num_params_set > 1:
+        raise ValueError("Only one of tolerance, rate, or precision can be set")
+
+    # Setup zfp structs to begin compression
+    cdef zfp_field* field = _init_field(arr)
+    stream = zfp_stream_open(NULL)
+
+    cdef zfp_type ztype = zfp_type_none;
+    cdef int ndim = arr.ndim;
+    if tolerance > -1:
+        zfp_stream_set_accuracy(stream, tolerance)
```

Comment:
This still allows for potential user error, e.g., if the user passes -0.5 as the tolerance.  tolerance, rate, and precision all have to be non-negative.

Reply:
:+1: great point!  Updated to check for `>= 0`